### PR TITLE
policy: Report reason inputs are nonstandard from AreInputsStandard

### PR DESCRIPTION
**Pull request by Empact**
This results in a more informative debug string to the validation state in the
case of failure.

### CODE REVIEW

File: src/policy/policy.cpp [hunk 1]
```diff
@@ -169,19 +170,28 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
...
             // convert the scriptSig into a stack, so we can inspect the redeemScript
-            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE))
+            ScriptError serror;
+            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE, &serror)) {
+                reason = strprintf("input %u script error: %s", i, ScriptErrorString(serror));
                 return false;
-            if (stack.empty())
+            }
+            if (stack.empty()) {
+                reason = strprintf("input %u script stack empty", i);
```
Comment (promag) [inline src/policy/policy.cpp]: Not tested?

File: src/policy/policy.h [hunk 2]
```diff
@@ -88,9 +88,10 @@ bool IsStandardTx(const CTransaction& tx, std::string& reason);
     /**
      * Check for standard transaction types
      * @param[in] mapInputs    Map of previous transactions that have outputs we're spending
+     * @param[out] reason      If return is false, contains the reason the inputs were judged non-standard
```
Comment (promag) [inline src/policy/policy.h]: `contains the reason of the first non-standard input`?

File: src/policy/policy.cpp [hunk 3]
```diff
@@ -169,19 +170,28 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
...
+        }
 
         if (whichType == TX_SCRIPTHASH)
         {
             std::vector<std::vector<unsigned char> > stack;
             // convert the scriptSig into a stack, so we can inspect the redeemScript
-            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE))
+            ScriptError serror;
+            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE, &serror)) {
+                reason = strprintf("input %u script error: %s", i, ScriptErrorString(serror));
```
Comment (promag) [inline src/policy/policy.cpp]: Not tested?

File: src/bench/ccoins_caching.cpp [hunk 4]
```diff
@@ -77,8 +77,10 @@ static void CCoinsCaching(benchmark::State& state)
 
     // Benchmark.
     while (state.KeepRunning()) {
-        bool success = AreInputsStandard(t1, coins);
-        assert(success);
+        std::string reason;
+        if (!AreInputsStandard(t1, coins, reason)) {
+            throw std::runtime_error(reason);
```
Comment (promag) [inline src/bench/ccoins_caching.cpp]: Why this change? It must succeed no?

**Review by promag — COMMENTED** [review]
Concept ACK.

**Comment by Empact** [comment]
@promag Addressed your comments, and after a fair amount of banging my head against the wall, managed to test all cases.

**Comment by sipa** [comment]
utACK 7298231aa8faaff65c261eac557448c223b76500

**Comment by luke-jr** [comment]
This is already part of #7533 

At the very least, please try to use the short rejection reasons therein...

**Comment by luke-jr** [comment]
https://github.com/bitcoin/bitcoin/pull/7533/commits/c59e7166253e962043b5fdccd5de8a21969d21e0

**Comment by Empact** [comment]
Thanks @luke-jr, I adapted the short reasons from c59e716. How does that look? Is it alright to put the input index in the rejection reason or better to append?

**Comment by Empact** [comment]
Also moved the string instantiation out of the bench loop.

**Comment by luke-jr** [comment]
Might be better to have the index at the end, not sure.

**Comment by Empact** [comment]
@luke-jr moved the input index to the end. Now I'm questioning the need for the `bad-txns-` prefix, given that these are to become debug output, rather than the reason string, which is what usually bears this prefix.

**Comment by luke-jr** [comment]
Why do you want to limit it to debug output? Reasons should be in the reason field.

**Comment by Empact** [comment]
@luke-jr Good point - split the reason and debug fields, and replaced `bad-txns-nonstandard-inputs` with one of: `bad-txns-script-nonstandard`, `bad-txns-scriptsig-error`, or `bad-txns-scriptcheck-sigops`.

**Comment by luke-jr** [comment]
Prefer to use the existing `bad-txns-input-script-unknown`, `bad-txns-input-scriptsig-failure`, `bad-txns-input-scriptcheck-missing`, and `bad-txns-input-scriptcheck-sigops`

**Comment by Empact** [comment]
Rebased for  #13429

**Comment by Empact** [comment]
Fixed rebase for #13429, dropped refactor commit, and made use of every reason recommended by @luke-jr

File: src/test/script_p2sh_tests.cpp [hunk 5]
```diff
@@ -362,8 +378,47 @@ BOOST_AUTO_TEST_CASE(AreInputsStandard)
...
+    static const unsigned char op_return[] = { OP_RETURN }; // SCRIPT_ERR_OP_RETURN
+
+    CMutableTransaction txToNonStd4;
+    txToNonStd4.vout.resize(1);
+    txToNonStd4.vout[0].scriptPubKey = GetScriptForDestination(key[1].GetPubKey().GetID());
+    txToNonStd4.vout[0].nValue = 1000;
+    txToNonStd4.vin.resize(1);
+    txToNonStd4.vin[0].prevout.n = 8;
+    txToNonStd4.vin[0].prevout.hash = txFrom.GetHash();
+    txToNonStd4.vin[0].scriptSig = CScript(&op_return[0], &op_return[sizeof(op_return)]);
```
Comment (practicalswift) [inline src/test/script_p2sh_tests.cpp]: Isn't `&op_return[sizeof(op_return)]` an array overrun?

File: src/test/script_tests.cpp [hunk 6]
```diff
@@ -1032,6 +1032,16 @@ BOOST_AUTO_TEST_CASE(script_PushData)
     BOOST_CHECK_MESSAGE(err == SCRIPT_ERR_OK, ScriptErrorString(err));
 }
 
+BOOST_AUTO_TEST_CASE(script_EvalScript_error)
+{
+    static const unsigned char op_return[] = { OP_RETURN }; // SCRIPT_ERR_OP_RETURN
+
+    ScriptError err;
+    std::vector<std::vector<unsigned char>> op_return_stack;
+    BOOST_CHECK(!EvalScript(op_return_stack, CScript(&op_return[0], &op_return[sizeof(op_return)]), SCRIPT_VERIFY_P2SH, BaseSignatureChecker(), SigVersion::BASE, &err));
```
Comment (practicalswift) [inline src/test/script_tests.cpp]: Same here?

**Comment by Empact** [comment]
@practicalswift replaced uses of `&array[sizeof(array]` with `array + sizeof(array)`

**Comment by Empact** [comment]
Stopped shadowing `reason` var (by reusing the existing var).
Limited test array access changes to only affecting the added tests.

**Comment by Empact** [comment]
Added code-blocks to tests to protect against accidental reuse, e.g. of `txToNonStd1` in other test cases.

**Comment by sipa** [comment]
utACK 85e2ddb14c15b9613093ec0cb991b77091c4287c, with or without nits addressed.

**Comment by Empact** [comment]
Updated:
* `bad-txns-input-scriptsig-failure` -> `bad-txns-input-p2sh-scriptsig-malformed`
* `bad-txns-input-scriptcheck-missing` -> `bad-txns-input-p2sh-no-redeemscript`
* `bad-txns-input-scriptcheck-sigops` -> `bad-txns-input-p2sh-redeemscript-sigops`

File: src/policy/policy.cpp [hunk 7]
```diff
@@ -169,16 +169,28 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
...
+            reason = "bad-txns-input-script-nonstandard";
+            debug = strprintf("input %u", i);
             return false;
         } else if (whichType == TX_SCRIPTHASH) {
             std::vector<std::vector<unsigned char> > stack;
             // convert the scriptSig into a stack, so we can inspect the redeemScript
-            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE))
+            ScriptError serror;
+            if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE, &serror)) {
+                reason = "bad-txns-input-scriptsig-failure";
```
[the remark this answers is not in the document]
Comment (luke-jr) [reply src/policy/policy.cpp]: The order of words typically indicates subcategories/relationships. For that reason, "bad-txns-scriptsig-p2sh-malformed" would fit better.

[the remark this answers is not in the document]
Comment (Empact) [reply src/policy/policy.cpp]: I’m currently using: “bad-txns-input-p2sh-scriptsig-malformed” for this.

Because all reasons in this function relate to inputs, of which are tested/subdivided according to type (p2sh in this case), then the subject scrptsig, then the issue.

Not that there’s a right or wrong here, but that’s the rationale.

File: src/validation.cpp [hunk 8]
```diff
@@ -706,8 +706,9 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
         }
 
         // Check for non-standard pay-to-script-hash in inputs
-        if (fRequireStandard && !AreInputsStandard(tx, view))
-            return state.Invalid(ValidationInvalidReason::TX_NOT_STANDARD, false, REJECT_NONSTANDARD, "bad-txns-nonstandard-inputs");
+        std::string debug;
+        if (fRequireStandard && !AreInputsStandard(tx, view, reason, debug))
```
Comment (MarcoFalke) [inline src/validation.cpp]: style-nit: Could add `{`?

**Review by maflcko — COMMENTED** [review]
Concept ACK

File: src/validation.cpp (hunk 8 above)
Comment (Empact) [reply src/validation.cpp]: Added 👍

**Comment by Empact** [comment]
Rebased, added missing braces.

**Comment by Empact** [comment]
Rebased

**Comment by sipa** [comment]
utACK 6a241604fec4346d90495fd468f54a6bf6a0ee88

**Comment by Empact** [comment]
Rebased for #18859

**Comment by 0xB10C** [comment]
Concept ACK

**Review by glozow — COMMENTED** [review]
Concept ACK to more informative error messages

**Comment by theStack** [comment]
Concept ACK

File: src/policy/policy.cpp [hunk 9]
```diff
@@ -186,20 +187,36 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
...
+            debug = strprintf("input %u", i);
+            return false;
+        } else if (whichType == TxoutType::WITNESS_UNKNOWN) {
             // WITNESS_UNKNOWN failures are typically also caught with a policy
             // flag in the script interpreter, but it can be helpful to catch
             // this type of NONSTANDARD transaction earlier in transaction
             // validation.
+            reason = "bad-txns-input-witness-unknown";
+            debug = strprintf("input %u", i);
             return false;
```
Comment (theStack) [inline src/policy/policy.cpp]: This if-branch is currently dead code, since the type `WITNESS_UNKNOWN` is already matched in the first if-condition a few lines above (obvioulsy, it was forgotten to be removed there). I.e. we want

```diff
diff --git a/src/policy/policy.cpp b/src/policy/policy.cpp
index 08f41376a..581fdcce3 100644
--- a/src/policy/policy.cpp
+++ b/src/policy/policy.cpp
@@ -186,7 +186,7 @@ std::optional<std::pair<std::string, std::string>> AreInputsStandard(const CTran

         std::vector<std::vector<unsigned char> > vSolutions;
         TxoutType whichType = Solver(prev.scriptPubKey, vSolutions);
-        if (whichType == TxoutType::NONSTANDARD || whichType == TxoutType::WITNESS_UNKNOWN) {
+        if (whichType == TxoutType::NONSTANDARD) {
             return std::make_pair(
                 "bad-txns-input-script-nonstandard",
                 strprintf("input %u", i)

```

To avoid issues like this, should probably also add a test for "bad-txns-input-witness-unknown"?

Comment (Empact) [reply src/policy/policy.cpp]: Good point! Updated here: https://github.com/bitcoin/bitcoin/compare/631d3c980090343f2f0d9dd97dac585c7fa94f13..1f3c017cce798bdcee1d719bf160dc07783239f2

**Review by theStack — COMMENTED** [review]
CI fails, seems like some functional tests need to be adapted to match the more specific failure reason string (https://cirrus-ci.com/task/6048174287618048?logs=ci#L3337):
```
...
AssertionError: [node 1] Expected messages "['bad-txns-input-script-nonstandard']" does not partially match log:
...
```

**Comment by Empact** [comment]
> CI fails, seems like some functional tests need to be adapted to match the more specific failure reason string (https://cirrus-ci.com/task/6048174287618048?logs=ci#L3337):
> 
> ```
> ...
> AssertionError: [node 1] Expected messages "['bad-txns-input-script-nonstandard']" does not partially match log:
> ...
> ```

Fixed: https://github.com/bitcoin/bitcoin/compare/1f3c017cce798bdcee1d719bf160dc07783239f2..dfa194343fde1557420a3b4a52b6a93e42adb848

File: src/test/script_p2sh_tests.cpp [hunk 10]
```diff
@@ -368,8 +390,63 @@ BOOST_AUTO_TEST_CASE(AreInputsStandard)
...
+
+    // TxoutType::WITNESS_UNKNOWN
+    CMutableTransaction txWitnessUnknown;
+    txWitnessUnknown.vout.resize(1);
+    txWitnessUnknown.vout[0].scriptPubKey = GetScriptForDestination(PKHash(key[1].GetPubKey()));
+    txWitnessUnknown.vout[0].nValue = 1000;
+    txWitnessUnknown.vin.resize(1);
+    txWitnessUnknown.vin[0].prevout.n = 9;
+    txWitnessUnknown.vin[0].prevout.hash = txFrom.GetHash();
+    txWitnessUnknown.vin[0].scriptSig << std::vector<unsigned char>(witnessUnknown.begin(), witnessUnknown.end());
```
Comment (achow101) [inline src/test/script_p2sh_tests.cpp:445]: In ba538fba899482f5549a6377d657d9d89b13bfd5 "Propagate validation debug information from AreInputsStandard"

nit: setting `scriptSig` is unnecessary as the thing being checked is the scriptPubKey of the output being spent.

**Review by achow101 — COMMENTED** [review]
The more descriptive reason strings seem to have gotten lost in the past couple of years.

**Review by aureleoules — COMMENTED** [review]
Instead of returning an option would the Result class make more sense?
Also, instead of a pair of strings, maybe create a new struct?

**Comment by achow101** [comment]
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.
